controllers/promise_controller.go

```diff
@@ -600,31 +696,53 @@ func (r *PromiseReconciler) applyWorkResourceForDependencies(ctx context.Context
 		return err
 	}
 
-	logger.Info("resource reconciled", "operation", op, "namespace", work.GetNamespace(), "name", work.GetName(), "gvk", work.GroupVersionKind())
+	o.logger.Info("resource reconciled", "operation", op, "namespace", work.GetNamespace(), "name", work.GetName(), "gvk", work.GroupVersionKind())
 	return nil
 }
 
-func (r *PromiseReconciler) generatePipelines(promise *v1alpha1.Promise, logger logr.Logger) ([]v1alpha1.Pipeline, []v1alpha1.Pipeline, error) {
-	var configurePipelines []v1alpha1.Pipeline
+type promisePipelines struct {
+	DeleteResource    []v1alpha1.Pipeline
+	ConfigureResource []v1alpha1.Pipeline
+	ConfigurePromise  []v1alpha1.Pipeline
+}
 
+func (r *PromiseReconciler) generatePipelines(promise *v1alpha1.Promise, logger logr.Logger) (promisePipelines, error) {
+	var configureResourcePipelines []v1alpha1.Pipeline
 	for _, pipeline := range promise.Spec.Workflows.Resource.Configure {
 		p, err := generatePipeline(pipeline, logger)
 		if err != nil {
-			return nil, nil, err
+			return promisePipelines{}, err
+		}
+		configureResourcePipelines = append(configureResourcePipelines, p)
+	}
+
+	var configurePromisePipelines []v1alpha1.Pipeline
+	for _, pipeline := range promise.Spec.Workflows.Promise.Configure {
+		p, err := generatePipeline(pipeline, logger)
+		if err != nil {
+			return promisePipelines{}, err
 		}
-		configurePipelines = append(configurePipelines, p)
+		configurePromisePipelines = append(configurePromisePipelines, p)
 	}
 
-	var deletePipelines []v1alpha1.Pipeline
+	var deleteResourcePipelines []v1alpha1.Pipeline
 	for _, pipeline := range promise.Spec.Workflows.Resource.Delete {
 		p, err := generatePipeline(pipeline, logger)
 		if err != nil {
-			return nil, nil, err
+			return promisePipelines{}, err
 		}
-		deletePipelines = append(deletePipelines, p)
+		deleteResourcePipelines = append(deleteResourcePipelines, p)
 	}
 
```

Comment:
could `generatePipeline` receive the unstructured array and return the structured pipeline? we kinda have the same for three times here.

Reply:
that is what it does 🤔 I will see if anything can be done to reduce the repetition